src/BoundaryConditions/flat_extrapolation_open_boundary_matching_scheme.jl

```diff
@@ -0,0 +1,152 @@
+using Oceananigans.Grids: xspacing, yspacing, zspacing
+
+"""
+    FlatExtrapolation
+
+Zero gradient perpendicular velocity boundary condition.
+
+We find the boundary value by Taylor expanding the gradient at the boundary point (`xᵢ`)
+to second order:
+```math
+f′(xᵢ) ≈ f′(xᵢ₋₁) + f′′(xᵢ₋₁)(xᵢ₋₁ - xᵢ) + O(Δx²) = f′(xᵢ₋₁) + f′′(xᵢ₋₁)Δx + O(Δx²),
+```
+where ``Δx=xᵢ₋₁ - xᵢ`` (for simplicity, we will also assume the spacing is constant at
+all ``i`` for now).
+We can substitute the gradinet at some point ``j`` (``f′(xⱼ)``) with the central 
+difference approximation:
+```math
+f′(xⱼ) ≈ (f(xⱼ₊₁) - f(xⱼ₋₁)) / 2Δx,
+```
+and the second derivative at some point ``j`` (``f′′(xⱼ)``) can be approximated as:
+```math
+f′′(xⱼ) ≈ (f′(xⱼ₊₁) - f′(xⱼ₋₁)) / 2Δx = ((f(xⱼ₊₂) - f(xⱼ)) - (f(xⱼ) - f(xⱼ₋₂))) / (2Δx)².
+```
+When we then substitute for the boundary adjacent point ``f′′(xᵢ₋₁)`` we know that 
+``f′(xⱼ₊₁)=f′(xᵢ)=0`` so the Taylor expansion becomes:
+```math
+f(xᵢ) ≈ f(xᵢ₋₂) - (f(xᵢ₋₁) - f(xᵢ₋₃))/2 + O(Δx²).
+```
+
+When the grid spacing is not constant the above can be repeated resulting in the factor 
+of 1/2 changes to ``Δx₋₁/(Δx₋₂ + Δx₋₃)`` instead, i.e.:
+```math
+f(xᵢ) ≈ f(xᵢ₋₂) - (f(xᵢ₋₁) - f(xᵢ₋₃))Δxᵢ₋₁/(Δxᵢ₋₂ + Δxᵢ₋₃) + O(Δx²)
+```.
+"""
+struct FlatExtrapolation{FT}
+    relaxation_timescale :: FT
+end
+
+const FEOBC = BoundaryCondition{<:Open{<:FlatExtrapolation}}
+
+function FlatExtrapolationOpenBoundaryCondition(val = nothing; relaxation_timescale = Inf, kwargs...)
+    classification = Open(FlatExtrapolation(relaxation_timescale))
+    
+    return BoundaryCondition(classification, val; kwargs...)
+end
+
+@inline function relax(l, m, ϕ, bc, grid, clock, model_fields)
```

Comment:
```suggestion
@inline function relax(l, m, grid, ϕ, bc, clock, model_fields)
```

this is definitely nitpicky but we may thank ourselves. Functions like this we typically write in "kernel form", which is either `i, j, k, grid` or `i, j, grid`. There are special advantages for diagnostics for `i, j, k, grid` and theoretically we may have a similar advantage for the two index form `i, j, grid` for reduced diagnostics (but don't right now).